File: backend/app/models/invoice.py
from decimal import Decimal
from enum import Enum
from sqlalchemy import (
    DECIMAL,
    BigInteger,
    Column,
    Date,
    ForeignKey,
    Integer,
    String,
    Index,
    Enum as SQLEnum,
    func
)
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import relationship

from ..database import Base


class InvoiceStatusEnum(str, Enum):
    PAID = "PAID"
    UNPAID = "UNPAID"
    OVERDUE = "OVERDUE"


class ClientTypeEnum(str, Enum):
    INDIVIDUAL = "INDIVIDUAL"
    BUSINESS = "BUSINESS"


class InvoiceTypeEnum(str, Enum):
    SERVICE = "SERVICE"
    PRODUCT = "PRODUCT"


class InvoiceItem(Base):
    __tablename__ = "invoice_items"
    __table_args__ = (
        Index('ix_invoice_items_invoice_id', 'invoice_id'),
    )

    id = Column(BigInteger, primary_key=True, index=True)
    invoice_id = Column(Integer, ForeignKey("invoices.id"), index=True)
    description = Column(String(200), nullable=False)
    quantity = Column(DECIMAL(10, 2), nullable=False, default=Decimal('0.00'))
    unit_price = Column(DECIMAL(10, 2), nullable=False, default=Decimal('0.00'))
    discount_percentage = Column(DECIMAL(5, 2), nullable=True, default=Decimal('0.00'))
    order = Column(Integer, nullable=False)

    invoice = relationship("Invoice", back_populates="items")
    subitems = relationship(
        "InvoiceSubItem",
        back_populates="invoice_item",
        cascade="all, delete-orphan",
        order_by="InvoiceSubItem.order"
    )

    @hybrid_property
    def line_total(self):
        return self.quantity * self.unit_price * (Decimal('1') - self.discount_percentage / Decimal('100'))


class InvoiceSubItem(Base):
    __tablename__ = "invoice_subitems"

    id = Column(BigInteger, primary_key=True, index=True)
    invoice_item_id = Column(BigInteger, ForeignKey("invoice_items.id"), index=True)
    description = Column(String(200), nullable=False)
    order = Column(Integer, nullable=False)

    invoice_item = relationship("InvoiceItem", back_populates="subitems")


class Invoice(Base):
    __tablename__ = "invoices"
    __table_args__ = (
        Index('ix_invoices_invoice_date', 'invoice_date'),
        Index('ix_invoices_total', 'total'),
        Index('ix_invoices_bill_to_id', 'bill_to_id'),
        Index('ix_invoices_send_to_id', 'send_to_id'),
        Index('ix_invoices_status', 'status'),
        Index('ix_invoices_client_type', 'client_type'),
        Index('ix_invoices_invoice_type', 'invoice_type'),
    )

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), index=True, nullable=False)
    invoice_number = Column(String(50), unique=True, index=True, nullable=False)
    invoice_date = Column(Date, nullable=False)
    bill_to_id = Column(Integer, ForeignKey("contacts.id"), index=True, nullable=False)
    send_to_id = Column(Integer, ForeignKey("contacts.id"), index=True, nullable=False)
    tax_rate = Column(DECIMAL(5, 2), nullable=False, default=Decimal('0.00'))
    discount_percentage = Column(DECIMAL(5, 2), nullable=True, default=Decimal('0.00'))
    notes = Column(String(500), nullable=True)
    template_id = Column(Integer, ForeignKey("templates.id"), nullable=False, index=True)

    status = Column(SQLEnum(InvoiceStatusEnum), nullable=False, default=InvoiceStatusEnum.UNPAID)
    client_type = Column(SQLEnum(ClientTypeEnum), nullable=False, default=ClientTypeEnum.INDIVIDUAL)
    invoice_type = Column(SQLEnum(InvoiceTypeEnum), nullable=False, default=InvoiceTypeEnum.SERVICE)

    subtotal = Column(DECIMAL(10, 2), nullable=False, default=Decimal('0.00'))
    tax = Column(DECIMAL(10, 2), nullable=False, default=Decimal('0.00'))
    total = Column(DECIMAL(10, 2), nullable=False, default=Decimal('0.00'))

    user = relationship("User", back_populates="invoices")
    bill_to = relationship("Contact", foreign_keys=[bill_to_id])
    send_to = relationship("Contact", foreign_keys=[send_to_id])
    items = relationship(
        "InvoiceItem",
        back_populates="invoice",
        cascade="all, delete-orphan",
        order_by="InvoiceItem.order"
    )
    template = relationship("Template")

    def calculate_totals(self):
        """Calculates and updates the subtotal, tax, and total for the invoice."""
        self.subtotal = sum(item.line_total for item in self.items)
        self.tax = self.discounted_subtotal * (self.tax_rate / Decimal('100'))
        self.total = self.discounted_subtotal + self.tax

    @property
    def discount_amount(self):
        return self.subtotal * (self.discount_percentage / Decimal('100'))
    
    @property
    def discounted_subtotal(self):
        return self.subtotal - self.discount_amount

File: frontend/src/types/invoice.ts
import { Contact, Template } from ".";

export interface InvoiceSubItemCreate {
    id?: number;
    description: string;
}

export interface InvoiceSubItem {
    id: number;
    description: string;
}

export interface InvoiceItemCreate {
    id?: number;
    description: string;
    quantity: number | null;
    unit_price: number | null;
    discount_percentage: number | null;
    subitems: InvoiceSubItemCreate[];
}

export interface InvoiceItem {
    id: number;
    invoice_id: number;
    description: string;
    quantity: number | null;
    unit_price: number | null;
    discount_percentage: number | null;
    subitems: InvoiceSubItem[];
    line_total: number;
}

export interface InvoiceCreate {
    invoice_number: string;
    invoice_date: string;
    bill_to_id: number | null;
    send_to_id: number | null;
    tax_rate: number;
    discount_percentage: number;
    notes?: string;
    items: InvoiceItemCreate[];
    template_id: number | null;
}

export interface Invoice {
    id: number;
    user_id: number;
    invoice_number: string;
    invoice_date: string;
    bill_to_id: number;
    bill_to_name: string;
    send_to_id: number;
    send_to_name: string;
    tax_rate: number;
    discount_percentage: number;
    notes?: string;
    items: InvoiceItem[];
    template_id: number;
    template_name: string;
    subtotal: number;
    tax: number;
    total: number;
    client_type?: string;
    invoice_type?: string;
    status?: string;
}

export interface GroupedInvoices {
    [key: string]: {
        invoices: Invoice[];
        invoice_count: number;
        total_amount: number;
    };
}

export interface InvoiceFilters {
    invoice_number?: string;
    bill_to_name?: string;
    send_to_name?: string;
    client_type?: string;
    invoice_type?: string;
    date_from?: string;
    date_to?: string;
    total_min?: string;
    total_max?: string;
    status?: string;
}

export interface InvoicePreviewProps {
    invoice: Invoice | InvoiceCreate;
    template: Template;
    billToContact?: Contact | null;
    sendToContact?: Contact | null;
}

export interface InvoiceItemFieldsProps {
    index: number;
    remove: (index: number) => void;
}

export interface InvoiceFormProps {
    contacts: Contact[];
    templates: Template[];
    isSubmitting: boolean;
    setSelectedTemplate: (template: Template | null) => void;
    isAuthenticated: boolean;
    handlePreview: () => void;
    isPDFGenerating: boolean;
    selectedTemplate: Template | null;
}

export type InvoiceFormErrors = {
    invoice_number?: string;
    invoice_date?: string;
    bill_to_id?: string;
    send_to_id?: string;
    tax_rate?: string;
    discount_percentage?: string;
    notes?: string;
    items?: {
        description?: string;
        quantity?: string;
        unit_price?: string;
        discount_percentage?: string;
        subitems?: {
            description?: string;
        }[];
    }[];
};

export interface InvoiceListResponse {
    items: Invoice[];
    total: number;
}

export interface InvoiceTotals {
    total_count: number;
    total_amount: number;
    status_counts: {
        [key: string]: number;
    };
}

export interface InvoiceListProps {
    invoices: Invoice[];
    contacts: Record<number, string>;
    templates: Record<number, string>;
    isMobile: boolean;
    onEdit: (id: number) => void;
    onDelete: (id: number) => void;
    onDownloadPDF: (invoice: Invoice) => void;
    sortBy: string;
    sortOrder: 'asc' | 'desc';
    onSort: (column: string) => void;
}